snitun/multiplexer/queue.py

```diff
@@ -0,0 +1,165 @@
+    queue: deque[MultiplexerMessage | None] = field(default_factory=deque)
+
+    @cached_property
+    def putters(self) -> deque[asyncio.Future[None]]:
+        """Return putters."""
+        return deque()
+
+
+class MultiplexerQueue:
+    """Multiplexer queue."""
+
+    def __init__(self, channel_size_limit: int) -> None:
+        """Initialize Multiplexer Queue.
+
+        Args:
+            channel_size_limit (int): The maximum size of a channel
+            data queue in bytes.
+
+        """
+        self._channel_size_limit = channel_size_limit
+        self._channels: defaultdict[MultiplexerChannelId, _ChannelQueue] = defaultdict(
+            _ChannelQueue,
+        )
+        # order controls which channel_id to get next
+        self._order: OrderedDict[MultiplexerChannelId, None] = OrderedDict()
+        self._getters: deque[asyncio.Future[None]] = deque()
+        self._loop = asyncio.get_running_loop()
+
+    def _wakeup_next(self, waiters: deque[asyncio.Future[None]]) -> None:
+        """Wake up the next waiter."""
+        while waiters:
+            waiter = waiters.popleft()
+            if not waiter.done():
+                waiter.set_result(None)
+                break
+
+    async def put(
+        self,
+        channel_id: MultiplexerChannelId,
+        message: MultiplexerMessage | None,
+    ) -> None:
+        """Put a message in the queue."""
+        # Based on asyncio.Queue.put()
+        channel = self._channels[channel_id]
+        while channel.total_bytes >= self._channel_size_limit:  # full
+            putter = self._loop.create_future()
+            channel.putters.append(putter)
+            try:
+                await putter
+            except:
+                putter.cancel()  # Just in case putter is not done yet.
+                with contextlib.suppress(ValueError):
+                    # Clean self._putters from canceled putters.
+                    channel.putters.remove(putter)
+                if not self.full(channel_id) and not putter.cancelled():
+                    # We were woken up by get_nowait(), but can't take
+                    # the call.  Wake up the next in line.
+                    self._wakeup_next(channel.putters)
+                raise
+        self._put(channel_id, channel, message)
+
+    def put_nowait(
+        self,
+        channel_id: MultiplexerChannelId,
+        message: MultiplexerMessage | None,
+    ) -> None:
+        """Put a message in the queue."""
+        self._put(channel_id, self._channels[channel_id], message)
+
+    def _put(
+        self,
+        channel_id: MultiplexerChannelId,
+        channel: _ChannelQueue,
+        message: MultiplexerMessage | None,
+    ) -> None:
+        """Put a message in the queue."""
+        size = 0 if message is None else len(message.data)
+        if channel.total_bytes >= self._channel_size_limit:
+            raise asyncio.QueueFull
+        channel.queue.append(message)
+        channel.total_bytes += size
```

Comment:
We can't get around counting total bytes as well because we need to it to know when to pause the reader. Likely when we get to 4*max size